### DISCUSSION TITLE
How to upgrade Heroku app?

### DISCUSSION THREAD

**Discussion by johnnypea** [babybuddy/babybuddy#418] [General]
What is the recommended way to upgrade on Heroku? Thanks.

**Comment by cdubz** [comment]
Hah... this has never come up before. What version are you on now? I assume you'll need to run some commands but I'd like to test before I advise you 😬

**Comment by johnnypea** [comment]
@cdubz I reinstalled recently so v1.10.1 🙏🏽

**Comment by cdubz** [comment]
@johnnypea surprisingly it seems like there is not actually a great way to do this... the recommended way to checkout the updated code, add your Heroku remote, and push the changes to it. Not quite as easy as the button they give you to start 🤔 

See e.g. [How To Update A Deployed App On Heroku](https://dev.to/jessesbyers/how-to-update-a-deployed-app-on-heroku-3fde)

**Comment by starbuck93** [comment]
I came across this today as I wanted to upgrade from 1.10 or something. I forked babybuddy into my profile and linked it to Heroku. Then on the Deploy page, /apps/baby-dean/deploy/github, I linked my github account and now it says "Automatic deploys from `master` are enabled". So all I have to do is "Fetch Upstream --> Fetch and Merge" in my fork when I want to update my instance.

**Reply by cdubz** [reply]
That's great to know! Still need to find somewhere to put this in docs...

**Reply by starbuck93** [reply]
Perhaps under the "Setup" tree (https://docs.baby-buddy.net/setup/upgrade)? But I also know I've done it a different way than the docs prescribed. I believe I started my instance with the Deploy to Heroku button, but then "moved" it to my fork.

**Reply by cdubz** [reply]
Hm yeah. I'd prefer to recommend the forking approach but I find it so annoying that GitHub Actions are enabled on forks by default that I hate ever recommending forking.

**Reply by starbuck93** [reply]
Well, seeing as Heroku is getting rid of free plans, looks like I'll be switching over to Docker..

**Reply by cdubz** [reply]
Yeah I'm pretty unhappy about that... for the demo and PR reviews and stuff I'm gonna look at self-hosting Dokku or Cap Rover.

I had a pretty decent experience with the Railway.app free tier. It wasn't sufficient for what the demo + PR reviewers but should be sufficient for a single instance if you're shoppin' around.

**Comment by whnr** [comment]
Just wanted to Tag on here. How do Upgrades work in general? Will the database just be compatible, are changes Managed in the background. Is everything compatible while on v1.x?

Background: I deployed the Heroku app while still in the Hospital. I have a personal server space at [uberspace](https://uberspace.de/en/) where I can run it. I want to migrate and upgrade

**Reply by cdubz** [reply]
Hm yeah I guess this really is not documented in general... let alone for Heroku specifically. The primary thing you'll need to do after any update is run `python manage.py migrate` within Baby Buddy's Python env (how you get there and run the commend will differ depending on the deployment). Most of the time there won't be any database migrations but they are needed on occasion.

**Reply by whnr** [reply]
I’ll try it and report / document as I go along.

**Reply by whnr** [reply]
Here's where I got (while the baby was sleeping):

### First try with a small diff

I found out that Heeroku just deploys whatever is on master and not the latest tag, so my first update was from v1.10.2 to a commit somewhere later. Here's how that worked super well:

Setup using uswgi, supervisorctl and mariadb (to which I migrated from the heroku postgres manualy—different story). I also have a branch called `local`.

One could follow the article [The Art Of Graceful Reloading](https://uwsgi-docs.readthedocs.io/en/latest/articles/TheArtOfGracefulReloading.html) but it's easier to tell the family to take notes for 10 minutes:

    supervisorctl stop uwsgi

Go to the babybuddy git folder. on the local git branch:

    git fetch 
    git merge <your tag here>

Resolve any merge confilct. E.g. might be in Pipfile since I had `mysqlclient` installed. Complete the merge:

    git commit

Update the apps dependencies and migrate the database.

    pipenv shell
    pipenv update
    export DJANGO_SETTINGS_MODULE=babybuddy.settings.production
    python manage.py migrate

Restart `uwsgi`:

    supervisorctl start uswgi

### Second try going from v1.10.2+ to v1.12.2

Some migration problems came up.

    Running migrations:
      No migrations to apply.
      Your models in app(s): 'babybuddy', 'core' have changes that are not yet reflected in a migration, and so won't be applied.
      Run 'manage.py makemigrations' to make new migrations, and then re-run 'manage.py migrate' to apply them.

I did as the message said (`python manage.py makemigrations`) and got:

````
Migrations for 'babybuddy':
  babybuddy/migrations/0023_alter_settings_timezone.py
    - Alter field timezone on settings
Migrations for 'core':
  core/migrations/0024_alter_tag_slug.py
    - Alter field slug on tag
````

Now I could successfully run `python manage.py migrate. Restart the server and :+1:
